google-cloud-contrib/google-cloud-nio/src/main/java/com/google/cloud/storage/contrib/nio/CloudStorageReadChannel.java

```diff
@@ -135,9 +140,29 @@ public int read(ByteBuffer dst) throws IOException {
     }
   }
 
+  private static boolean isReopenable(Throwable exs) {
+    Throwable throwable = exs;
+    // ensures finite iteration
+    int maxDepth = 10;
+    while (throwable != null && maxDepth-- > 0) {
+      Throwable cause = throwable.getCause();
+      if (throwable.getMessage().contains("Connection closed prematurely")
+          || cause instanceof SSLException
+          || cause instanceof EOFException
+          || cause instanceof SocketException
+          || cause instanceof SocketTimeoutException) {
+        return true;
+      }
+      throwable = cause;
+    }
+    return false;
+  }
+
   private void sleepForAttempt(int attempt) {
+    // exponential backoff, but let's bound it around 2min.
+    long delay = 1000L * (1L << Math.min(attempt, 7));
     try {
-      Thread.sleep((attempt - 1) * 500);
+      Thread.sleep(delay);
```

Comment:
It looks like the previous implementation would sleep 0 milliseconds for the first time (attempt - 1). That seems weird, so I like the new version more, but mentioning it here just in case.